[DNM] Test-only branch

msirek opened the pull request:
This branch is used only for running tests.
Nothing to review here.

Release note: None

Commit:
norm: early inlining of WITH bound expression with VALUES

Fixes https://github.com/cockroachlabs/support/issues/1769

This commit implements early inlining of WITH clause bound expressions
(at the time when the WithScanExpr is normalized) if the
WITH bound expression is a ValuesExpr with all constant expressions or
placeholders and it appears in an IN or NOT IN expression, for example:
`column IN (VALUES(...))` or `column NOT IN(VALUES(...))`.
Not doing early inlining can result in a superfluous join.

Normally WITH clause inlining is done when the top-level WithExpr is
normalized, but since normalization is done bottom-up, the WithExpr is
processed last, making the inlining too late to trigger other
normalization rules.

Release note (performance improvement): This patch adds early inlining
of VALUES clauses and unnested arrays in WITH queries in order to
eliminate unnecessary joins.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/88920)
<!-- Reviewable:end -->

cucaroach commented:
I think if you mark this as draft no one will be notifications about it.

msirek commented:
> I think if you mark this as draft no one will get notifications about it.

Thanks!

Commit:
opt: add WithBindingInfo to metadata

This commit adds additional information regarding WITH bound expressions
to the withBindings map in order to avoid increasing memo memory use.

Release note: None
